### PR TITLE
Play next/previous video in folder

### PR DESCRIPTION
**Pull request by tbertels**
By default, PgDown and PgUp will play the next/previous video in folder if there's no more chapter and there's only one file in the playlist.

With the renamed "Play next file in folder when there is only one item in the playlist" option enabled, the next file in folder will play without pressing anything when the end of the file is reached and the playlist has only the current file.

Fixes issue #137

### CODE REVIEW

**Comment by tbertels** [comment]
`PlaybackManager::checkAfterPlayback(bool playlistMode)` is always called with `playlistMode` set to `true`.
It used to be also called with `false`.
Could it be removed now?

**Comment by tbertels** [comment]
I'm thinking about adding key shortcuts to open the next/previous file in folder even if there are more chapters after, I think MPC-HC has this.

File: manager.cpp [hunk 1]
```diff
@@ -248,7 +248,7 @@ void PlaybackManager::navigateToNextChapter()
 {
     int64_t nextChapter = mpvObject_->chapter() + 1;
     if (nextChapter >= numChapters)
-        mpvw_playbackIdling();
+        playNext();
```
Comment (cmdrkotori) [inline manager.cpp:251]: This change skips the check for decrementing the extra play times counter when navigating to past-the-end chapter.  If you wish to keep this, the extra play times counter needs to be set back to zero.

**Comment by cmdrkotori** [comment]
> `PlaybackManager::checkAfterPlayback(bool playlistMode)` is always called with `playlistMode` set to `true`. It used to be also called with `false`. Now it seems like we're always in playlist mode, it just contains one file or more. Could it be removed?

Yes it could be removed.  There used to be some obscure mode where a file could be played without it landing in the playlist, but it's gone now.

> I'm thinking about adding key shortcuts to open the next/previous file in folder even if there are more chapters after, I think MPC-HC has this.

If there's something MPC-HC is doing that we aren't, I consider it a bug unless there's a good reason why we can't/don't.

> Maybe "Play next file in folder when there is only one item in the playlist" could be renamed to something like "Play automatically next file in folder when there is only one item in the playlist" to make it clearer that the next file will play... automatically.

I'm not against this. It is also possible to add a separate option (default on) for controlling whether automatic file progression is active or not.

File: manager.cpp [hunk 2]
```diff
@@ -646,7 +646,7 @@ void PlaybackManager::updateSubtitleTrack()
     mpvObject_->setSubtitleTrack(subtitleEnabled ? subtitleTrackSelected : 0);
 }
 
-void PlaybackManager::checkAfterPlayback(bool playlistMode)
+void PlaybackManager::checkAfterPlayback()
```
Comment (cmdrkotori) [inline manager.cpp:649]:
```
manager.cpp:820:27: error: no matching function for call to ‘PlaybackManager::checkAfterPlayback(bool)’
  820 |         checkAfterPlayback(true);
      |         ~~~~~~~~~~~~~~~~~~^~~~~~
```
Might want to amend this commit + force-push.

File: manager.cpp (hunk 1 above)
Comment (cmdrkotori) [reply manager.cpp:253]: Yes and yes.  For testing purposes each track can also be changed individually by the "Extra Play Times" submenu under "Edit".
And you're right, I forgot about that setting.

**Comment by tbertels** [comment]
> > I'm thinking about adding key shortcuts to open the next/previous file in folder even if there are more chapters after, I think MPC-HC has this.
> 
> If there's something MPC-HC is doing that we aren't, I consider it a bug unless there's a good reason why we can't/don't.
> 
I've just checked it out: it opens the next/previous file in playlist or, if there's only one, the next file in folder with Ctrl+PgDown and Ctrl+PgUp by default. 

> > Maybe "Play next file in folder when there is only one item in the playlist" could be renamed to something like "Play automatically next file in folder when there is only one item in the playlist" to make it clearer that the next file will play... automatically.
> 
> I'm not against this. It is also possible to add a separate option (default on) for controlling whether automatic file progression is active or not.

By "automatic file progression", do you mean playing the next file in folder without user interaction?
Because MPC-HC doesn't do that, you have to press PgDown.

So I was thinking about an option to open the next file in folder manually (enabled by default) and an option to open it automatically (disabled by default).
Maybe "Use next file in folder when there is only one item in the playlist" and "Play automatically next file in folder when there is only one item in the playlist". I'm not sure that's clear enough.

**Comment by cmdrkotori** [comment]
> By "automatic file progression", do you mean playing the next file in folder without user interaction?
> Because MPC-HC doesn't do that, you have to press PgDown.

I was speaking in general (both playlist and folder) in relation to the existing code.  But you're right MPC-HC doesn't do that, it should be changed to match.  Playlists without automatic playlist progression doesn't make sense anyway.

>So I was thinking about an option to open the next file in folder manually (enabled by default) and an option to open it automatically (disabled by default).
>
>Maybe "Use next file in folder when there is only one item in the playlist" and "Play automatically next file in folder when there is only one item in the playlist". I'm not sure that's clear enough though.

It's clear enough. You could also go with "Use next file in folder when there is only one item in the playlist" (enabled by default, as you said) and "Wait for navigation when there is only one item in the playlist" (enabled by default).  I'll leave this up to you.

**Comment by tbertels** [comment]
> > So I was thinking about an option to open the next file in folder manually (enabled by default) and an option to open it automatically (disabled by default).
> > Maybe "Use next file in folder when there is only one item in the playlist" and "Play automatically next file in folder when there is only one item in the playlist". I'm not sure that's clear enough though.
> 
> It's clear enough. You could also go with "Use next file in folder when there is only one item in the playlist" (enabled by default, as you said) and "Wait for navigation when there is only one item in the playlist" (enabled by default). I'll leave this up to you.

MPC-HC actually uses the equivalent of this Playback option we talked about in the issue to enable or not the auto play next file in folder:

image

And it does it this way: first the files of the playlist, then the files in the folder.

MPC-QT does the same with the equivalent setting.
So I wonder if it wouldn't be better to just use that option instead of adding a second one which will do the exact same thing, now that the previous/next command feature will have its own option.

**Comment by cmdrkotori** [comment]
>So I wonder if it wouldn't be better to just use that option instead of adding a second one which will do the exact same thing, now that the previous/next command feature will have its own option.

Yes it would indeed. In fact I noticed when writing the code for "default after playback" (added much later) that I was stomping on my own foot having a semi-related option in the tweaks page, so I had to write it such a way that certain actions were invoked by multiple options, certain options took priority, and so on.

**Comment by cmdrkotori** [comment]
>I don't know if the users really want this though, as it can make them lose the playlist.

This is a concern, yes. I know of at least one mpv dev who uses mpc-qt to play their 10000-file-strong music collection that I presume is spread over multiple folders, so this is probably not desirable.  Adding so many files back would be a pain for them.

>If that's a real use case, I think it would be better to add a command to empty the playlist, which would allow continuing playing the files in the folder.

To support the existing use-case, there would have to be an option to turn it off I think.

**Comment by tbertels** [comment]
An option to turn off the command?
Given the fact that the user can just open a different file to empty the playlist, maybe it's just better to let it that like it is right now and just rename the option "Play next in the folder" to let the user know that it works just like the next file and next chapter (once at the end) commands (playlist if not empty, otherwise folder).

**Comment by cmdrkotori** [comment]
Yeah I'm fine with that.

File: mpvwidget.h [hunk 3]
```diff
@@ -76,6 +76,7 @@ class MpvObject : public QObject
     void setVideoTrack(int64_t id);
     void setDrawLogo(bool yes);
     void setVolume(int64_t volume);
+    bool eofReached();
```
Comment (cmdrkotori) [inline mpvwidget.h:79]: nitpick: this should be further down the page with the other getters, perhaps just before playLength().

**Review by cmdrkotori — APPROVED** [review]
LGTM. I'll do some testing and merge it in the morning if that's okay with you.

**Pull request merged**